File: backend/app/models/product.rb
class Product < ApplicationRecord
  validates :name, :price, :expiration_date, presence: true
  validates :price, numericality: { greater_than: 0 }
  validate :valid_name_format

  belongs_to :foreign_exchange

  private

  def valid_name_format
    return if name.blank?
    return if name.match?(/^[a-zA-Z0-9]+(?:[\s-][a-zA-Z0-9]+)*\s*#\s*\(\d+\)$/)

    errors.add(:name,
               'wrong name')
  end
end

File: frontend/src/types/product.ts
export interface Rate {
  currency: string;
  description: string;
  rate: number;
}

export interface ForeignExchange {
  date: string;
  rates: Rate[];
}

export interface Product {
  name: string;
  price: number;
  expiration_date: string;
  foreign_exchange?: ForeignExchange;
} 